### ISSUE TITLE
Error with rule get_all_bins, [Errno 17] File exists

### ISSUE THREAD

**Issue by mmanno** [metagenome-atlas/atlas#290]
Hello,

I am trying to launch ATLAS on 6 metagenomes samples (70 millions of paired-reads each) on a SLURM cluster.
I have an issue with the rule "get_all_bins" and the fallowing error :

```
Error in rule get_all_bins:
    jobid: 0
    output: genomes/all_bins

RuleException:
FileExistsError in line 25 of /home/mmanno/.conda/envs/atlas_200304/lib/python3.6/site-packages/atlas/rules/genomes.smk:
[Errno 17] File exists: '../../2019-3924-70/binning/DASTool/bins/metabat.066.fasta' -> 'genomes/all_bins/metabat.066.fasta'
  File "/home/mmanno/.conda/envs/atlas_200304/lib/python3.6/site-packages/atlas/rules/genomes.smk", line 25, in __rule_get_all_bins
  File "/home/mmanno/.conda/envs/atlas_200304/lib/python3.6/site-packages/atlas/scripts/utils/io.py", line 53, in symlink_relative
  File "/home/mmanno/.conda/envs/atlas_200304/lib/python3.6/concurrent/futures/thread.py", line 56, in run
Exiting because a job execution failed. Look above for error message
```

Can you help me to solve this problem ?
It seem the creation of bins symbolic links failed, but I don't know how to fix it. Do you have any idea ?

here the binning conf : 

```
########################
# Binning
########################

final_binner: DASTool             # [DASTool or one of the binner, e.g. maxbin]

binner:                           # If DASTool is used as final_binner, use predictions of this binners
- metabat
- maxbin

```

Thanks in advance
Maxime

**Comment by SilasK** [comment]
Yes, atlas tried to create symbolic links. From the folder of each sample to the genomes_all bins file.

1. Try to remove the genomes/all_bins folder and restart.

Orherwise copy yourself the all the genomes and run atlas.

**Comment by mmanno** [comment]
Hi,

The genome folder do not contain the all_bins sub_folder. It seem's Atlas deleted the folder.

```
genomes/
├── checkm
│   ├── all_bins_markers.fasta
│   └── checkm_all_bins.tsv
└── clustering
    └── all_contigs2bins.tsv.gz
```

And many bins from different samples have the same name, Is it ok ? if yes, I do not understand how it is supposed to create the links. 

```
analysis_ATLAS_total $ find 2019-39* -name metabat.066.fasta
2019-3916-70/binning/DASTool/bins/metabat.066.fasta
2019-3918-70/binning/DASTool/bins/metabat.066.fasta
2019-3924-70/binning/DASTool/bins/metabat.066.fasta
```

Thanks for your help,
Maxime

**Comment by SilasK** [comment]
Hmm, there seems to be the problem. In theory the bin names should be unique e.g. DASTool/bins/{sample}_metabat_01.fasta

What could be the problem?

- Are you using the latest version of Atlas?
- could you send me the head of the following files for one of the samples:
"{sample}/binning/DASTool/cluster_attribution.tsv”
"{sample}/binning/metabat/cluster_attribution.tsv”

And the log file:
"{sample}/logs/binning/get_bins_DASTool.log"

**Comment by mmanno** [comment]
hi,

Here the Atlas version I use :
```
analysis_ATLAS_total $ atlas --version
atlas, version 2.3.0
```

The head of the attibution files : 
```
analysis_ATLAS_total $ head 2019-3916-70/binning/DASTool/cluster_attribution.tsv     2019-3916-70_0  maxbin.005
2019-3916-70_9  maxbin.005
2019-3916-70_11 maxbin.005
2019-3916-70_20 maxbin.005
2019-3916-70_39 maxbin.005
2019-3916-70_131        maxbin.005
2019-3916-70_355        maxbin.005
2019-3916-70_491        maxbin.005
2019-3916-70_1493       maxbin.005
2019-3916-70_11455      maxbin.005
head 2019-3916-70/binning/metabat/cluster_attribution.tsv
0       1
2019-3916-70_0  2019-3916-70_metabat_001
2019-3916-70_1  2019-3916-70_metabat_002
2019-3916-70_2  2019-3916-70_metabat_003
2019-3916-70_3  2019-3916-70_metabat_004
2019-3916-70_4  2019-3916-70_metabat_004
2019-3916-70_5  2019-3916-70_metabat_005
2019-3916-70_6  2019-3916-70_metabat_006
2019-3916-70_7  2019-3916-70_metabat_004
2019-3916-70_8  2019-3916-70_metabat_007
```
The problem should be in the DASTool part, as I can see...

And the get_bins_DASTool.log is empty.
Do you need an other log file ?

**Comment by SilasK** [comment]
Yes, it seems that DASTool don't like your sample names and renamed the bins so that they are no longer unique. 

Yes, please send me the "{sample}/logs/binning/DASTool.log" 

For the next time, it's probably better to use sample names without hyphen nor underline.
For this time you would need to rename the bins.

The easiest way is to modify the "{sample}/binning/metabat/cluster_attribution.tsv" for all the samples so that it respects my naming scheme. 

Then you can rerun atlas with the options `-R get_bins` and it will re-run checkm and combine the bins. 

What do you think.

**Comment by mmanno** [comment]
Here the DASTool log :
Usearch is not installed, so the script used Diamond
```
analysis_ATLAS_total $ cat 2019-3916-70/logs/binning/DASTool.log
diamond v0.9.30.131 (C) Max Planck Society for the Advancement of Science
Documentation, support and updates available at http://www.diamondsearch.org

Running DAS Tool using 16 threads.
diamond v0.9.30.131 (C) Max Planck Society for the Advancement of Science
Documentation, support and updates available at http://www.diamondsearch.org

identifying single copy genes using diamond version 0.9.30
WARNING: Some scaffold names of2019-3916-70/binning/DASTool/metabat.scaffolds2bindo not match assembly headers:
        0
WARNING: Some scaffold names of2019-3916-70/binning/DASTool/maxbin.scaffolds2bindo not match assembly headers:
        0
WARNING: non-unique bin-IDs found
        creating unique bin-IDs

evaluating bin-sets
starting bin selection from 165 bins
||||||||||||||||||||||||||||||||||||||||||||||||||||||||||
bin selection complete: 58 bins above score threshold selected.
```

I am going to do what you propose.
Thanks

**Comment by mmanno** [comment]
Hello,

I corrected the cluster_attribution.tsv file and re-ran Atlas, and it is done !
I am checking all the outputs to be sure, but it seem's to be ok.

Thanks for you help.
All the best
Maxime

**Comment by johanneswerner** [comment]
Thank you for the solution, I ran into the same problems. Do you have recommendations for the sample names to avoid this error with new runs? 

Usually I use either dashes or underscores in my sample names - should they be entirely avoided? Or replaced by "." or CamelCase? Thank you for your help.

**Comment by mmanno** [comment]
> 
> 
> Thank you for the solution, I ran into the same problems. Do you have recommendations for the sample names to avoid this error with new runs?
> 
> Usually I use either dashes or underscores in my sample names - should they be entirely avoided? Or replaced by "." or CamelCase? Thank you for your help.

Hi @johanneswerner ,
Apparently, the issue will be solved in a upcomming update #298 and have no relation to the specials carracters we can put into the samples names.

Bye
Maxime

**Comment by johanneswerner** [comment]
@mmanno Thank you very much. :-)
